### ISSUE TITLE
some questions to clarify

### ISSUE THREAD

**Issue by yoshum** [mods333/energy-based-scene-graph#2]
Thank you for your inspiring work!

Just to clarify, let me ask some questions.

First is about where you use the energy-based model. Do you use the energy-based model only in training or also in inference? Apparently, `energy_test_net.py` does not use `energy_model`, but I'm not perfectly sure.

Second is about training of the scene graph generation model through the energy loss. In the paper, you wrote "Differentiation through the optimization path guides parameter learning of the model M that generates G_SG^0". Does this mean that you optimize the parameters of M as well as the parameters \theta of the energy model to minimize the energy loss (+ reg loss + task loss)?

**Comment by hu-my** [comment]
I have the same problem. In energy_test_net.py, the author calls "inference" function, which doesn't use loaded energy model.
As for the second question, I think the author may only optimize the parameter \theta of the energy model, because only energy_optimizer is called during training phrase, and the parameter of base model is not  included in energy_optimizer. So if I understand correctly, this work is trained in a postprocessing-like fashion.

**Comment by mods333** [comment]
Yes, the energy model is not used during inference. You should think of the energy model as a generalized loss function that guides the learning of the prediction model. We tried using the energy model during inference as well but didn't see any significant difference in the result. It might help to think in terms of an adversarial network, where the discriminator is used for training but discarded during inference. 
For the second question, yes,  the parameters of M and the energy model are optimized to minimize the total loss.
https://github.com/mods333/energy-based-scene-graph/blob/master/tools/energy_joint_train_cd.py#L263-L283.

@hu-my I believe you are looking into energy_train_net.py. The training script for the model is in energy_joint_train_cd.py. Sorry for the confusion. I used energy trained net for some of the early experiments in the project and forgot to remove it before releasing the codebase.
